DigitalLearningSolutions.Data/Services/ActivityService.cs

```diff
@@ -126,6 +132,28 @@ public byte[] GetActivityDataFileForCentre(int centreId, ActivityFilterData filt
             return stream.ToArray();
         }
 
+        public (DateTime startDate, DateTime? endDate)? GetValidatedUsageStatsDateRange(
+            string startDateString,
+            string endDateString,
+            int centreId
+        )
+        {
+            var parsedStartDate = DateTime.Parse(startDateString);
+            var start = GetStartOfActivityForCentre(centreId);
+            if (parsedStartDate < GetStartOfActivityForCentre(centreId))
+            {
+                return null;
+            }
+
+            if (DateTime.TryParse(endDateString, out var parsedEndDate) &&
+                (parsedEndDate < parsedStartDate || parsedEndDate > DateTime.Now))
+            {
+                return null;
+            }
+
+            return (parsedStartDate, parsedEndDate > DateTime.MinValue ? (DateTime?)parsedEndDate : null);

```

Comment:
Do we need to check for minvalue if it has to be after start date anyway?

Reply:
It only checks if it is after the start date if the end date is set. TryParse returns DateTime.MinValue for invalid or empty strings. I think it'll be clearer if I move this check to right after the parse however.